File: cpp-server/src/room.cpp
#include "room.h"
#include "session.h"   // full definition of Session
#include <iostream>
#include <unordered_map>
#include <chrono>
using json = nlohmann::json;

Room::Room() : nextPlayerId(1), m_lastActivity(std::chrono::steady_clock::now()) {}

void Room::updateActivity() {
    m_lastActivity = std::chrono::steady_clock::now();
}

std::chrono::steady_clock::time_point Room::getLastActivity() const {
    return m_lastActivity;
}

// Default constructor is now defined in header

void Room::join(std::shared_ptr<Session> s, const std::string& username) {
    nlohmann::json joinMsg;
    bool isNewPlayer = false;

    {
        std::lock_guard<std::mutex> lock(m_mutex);

        if (players.find(username) == players.end()) {
            Player p{ nextPlayerId++, username, 0 };
            players[username] = p;
            isNewPlayer = true;

            joinMsg = {
                {"type", "join"},
                {"payload", {
                    {"id", p.id},
                    {"username", p.username}
                }}
            };
        }

        if (s) {
            m_sessions.insert(s);
        }
        
        // Update activity timestamp
        updateActivity();
    }

    // Broadcast outside of mutex lock to avoid deadlock
    if (isNewPlayer) {
        broadcast(joinMsg.dump());
    }

    if (s) {
        replayPlayers(s);
        replayHistory(s);
    }
}


bool Room::leave(std::shared_ptr<Session> s) {
    std::lock_guard<std::mutex> lock(m_mutex);

    // just remove the session
    auto it = m_sessions.find(s);
    if (it != m_sessions.end()) {
        m_sessions.erase(it);
    }

    return m_sessions.empty();
}

void Room::resetLobby() {
    std::lock_guard<std::mutex> lock(m_mutex);
    players.clear();
    nextPlayerId = 1;
}

void Room::broadcast(const std::string& msg) {
    for (auto& s : m_sessions) {
        if (s) s->send(msg);
    }
}

bool Room::empty() {
    std::lock_guard<std::mutex> lock(m_mutex);
    return m_sessions.empty();
}

void Room::endRound() {
    json msg = { {"type","round_end"}, {"payload","Round finished!"} };
    broadcast(msg.dump());
}

bool Room::hasPlayer(const std::string& username) {
    std::lock_guard<std::mutex> lock(m_mutex);
    return players.find(username) != players.end();
}

// room.cpp
void Room::addStroke(const json& stroke) {
    std::lock_guard<std::mutex> lock(m_mutex);
    strokeHistory.push_back(stroke);
    updateActivity();
}

void Room::clearHistory() {
    std::lock_guard<std::mutex> lock(m_mutex);
    strokeHistory.clear();
}

void Room::replayHistory(std::shared_ptr<Session> s) {
    std::vector<json> strokesCopy;

    {
        std::lock_guard<std::mutex> lock(m_mutex);
        strokesCopy = strokeHistory; // Copy the strokes
    }

    // Send strokes outside of mutex lock
    for (auto& stroke : strokesCopy) {
        std::cout << "[DEBUG] Replaying stroke to " << (s ? "session" : "null") << "\n";
        if (s) s->send(stroke.dump());
    }
}

void Room::replayPlayers(std::shared_ptr<Session> s) {
    std::vector<std::pair<int, std::string>> playerData;

    {
        std::lock_guard<std::mutex> lock(m_mutex);
        for (auto& [username, p] : players) {
            playerData.push_back({ p.id, p.username });
        }
    }

    // Send player data outside of mutex lock
    for (auto& [id, username] : playerData) {
        nlohmann::json joinMsg = {
            {"type", "join"},
            {"payload", {
                {"id", id},
                {"username", username}
            }}
        };
        if (s) s->send(joinMsg.dump());
    }
}

std::unordered_set<std::string> Room::getPlayerUsernames() const {
    std::lock_guard<std::mutex> lock(m_mutex);
    std::unordered_set<std::string> usernames;
    for (const auto& [username, p] : players) {
        usernames.insert(username);
    }
    return usernames;
}

File: game-server/src/room.ts
import type { ClientSession, Player, Round } from "./types.js";

export interface RoomChatAnnouncements {
  onRoundStart?: () => void;
  onCorrectGuess?: (username: string) => void;
  onRoundTimeout?: (word: string) => void;
  onRoundSkipped?: (word: string) => void;
}

export class Room {
  private sessions = new Set<ClientSession>();
  private players = new Map<string, Player>();
  private nextPlayerId = 1;
  private strokeHistory: Record<string, unknown>[] = [];
  private lastActivity = Date.now();
  private chatAnnouncements: RoomChatAnnouncements = {};
  private currentRound: Round = {
    word: "",
    hint: "",
    active: false,
    startTime: 0,
    duration: 60,
  };
  private roundTimer: ReturnType<typeof setTimeout> | null = null;

  private playerKey(username: string): string {
    return username.toLowerCase();
  }

  setChatAnnouncements(announcements: RoomChatAnnouncements): void {
    this.chatAnnouncements = announcements;
  }

  join(session: ClientSession | null, username: string): void {
    let isNewPlayer = false;
    let joinMsg: Record<string, unknown> | null = null;
    const key = this.playerKey(username);

    if (!this.players.has(key)) {
      const p: Player = { id: this.nextPlayerId++, username, score: 0 };
      this.players.set(key, p);
      isNewPlayer = true;
      joinMsg = {
        type: "join",
        payload: { id: p.id, username: p.username },
      };
      console.log(`[DEBUG] Room::join - Adding new player: ${username}`);
    }

    if (session) {
      this.sessions.add(session);
    }
    this.lastActivity = Date.now();

    if (isNewPlayer && joinMsg) {
      this.broadcast(JSON.stringify(joinMsg));
    }

    if (session) {
      this.replayPlayers(session);
      this.replayHistory(session);
    }
  }

  leave(session: ClientSession): boolean {
    this.sessions.delete(session);
    return this.sessions.size === 0;
  }

  resetLobby(): void {
    this.players.clear();
    this.nextPlayerId = 1;
  }

  broadcast(msg: string): void {
    for (const s of this.sessions) {
      s.send(msg);
    }
  }

  isEmpty(): boolean {
    return (
      this.sessions.size === 0 &&
      this.players.size === 0 &&
      this.strokeHistory.length === 0 &&
      !this.currentRound.active
    );
  }

  hasPlayer(username: string): boolean {
    return this.players.has(this.playerKey(username));
  }

  getPlayers(): Map<string, Player> {
    return this.players;
  }

  getPlayerUsernames(): string[] {
    return [...this.players.keys()];
  }

  getStrokeHistory(): Record<string, unknown>[] {
    return this.strokeHistory;
  }

  getCurrentRound(): Round {
    return this.currentRound;
  }

  getLastActivity(): number {
    return this.lastActivity;
  }

  addStroke(stroke: Record<string, unknown>): void {
    this.strokeHistory.push(stroke);
    this.lastActivity = Date.now();
  }

  clearHistory(): void {
    this.strokeHistory = [];
  }

  /** Remove the most recent drawn line (start → draw* → end). */
  undoLastStroke(): boolean {
    if (this.strokeHistory.length === 0) return false;

    while (this.strokeHistory.length > 0) {
      const stroke = this.strokeHistory[this.strokeHistory.length - 1];
      const payload = (stroke as { payload?: { action?: string } }).payload;
      this.strokeHistory.pop();
      if (payload?.action === "start") break;
    }

    this.lastActivity = Date.now();
    return true;
  }

  startRound(word: string): void {
    if (this.roundTimer) {
      clearTimeout(this.roundTimer);
      this.roundTimer = null;
    }

    this.currentRound = {
      word,
      hint: "_".repeat(word.length),
      active: true,
      startTime: Date.now(),
      duration: 60,
    };

    const msg = {
      type: "round_start",
      payload: {
        word: this.currentRound.word,
        hint: this.currentRound.hint,
        time: this.currentRound.duration,
      },
    };
    this.broadcast(JSON.stringify(msg));
    this.startServerTimer();
    this.chatAnnouncements.onRoundStart?.();
  }

  handleGuess(username: string, guess: string): void {
    if (!this.currentRound.active) {
      return;
    }

    if (!this.hasPlayer(username)) {
      return;
    }

    const normalizedGuess = guess.trim().toLowerCase();
    const normalizedWord = this.currentRound.word.trim().toLowerCase();

    if (normalizedGuess !== normalizedWord) {
      return;
    }

    const player = this.players.get(this.playerKey(username))!;
    player.score += 100;

    console.log(`[ROOM] ${username} guessed correctly`);

    this.chatAnnouncements.onCorrectGuess?.(username);

    this.broadcast(
      JSON.stringify({
        type: "guess",
        payload: {
          user: username,
          word: guess,
          correct: true,
          score: player.score,
        },
      }),
    );
    this.endRoundInternal();
  }

  endRound(): void {
    if (!this.currentRound.active) return;
    const word = this.currentRound.word;
    this.chatAnnouncements.onRoundTimeout?.(word);
    this.endRoundInternal();
  }

  skipRound(): void {
    if (!this.currentRound.active) return;
    const word = this.currentRound.word;
    this.chatAnnouncements.onRoundSkipped?.(word);
    this.endRoundInternal();
  }

  private endRoundInternal(): void {
    if (!this.currentRound.active) return;
    this.currentRound.active = false;

    if (this.roundTimer) {
      clearTimeout(this.roundTimer);
      this.roundTimer = null;
    }

    const scores: Record<string, number> = {};
    for (const [username, p] of this.players) {
      scores[username] = p.score;
    }

    this.broadcast(
      JSON.stringify({
        type: "round_end",
        payload: { word: this.currentRound.word, scores },
      }),
    );
  }

  private startServerTimer(): void {
    this.roundTimer = setTimeout(() => {
      console.log("[ROOM] Server timer expired - ending round");
      this.endRound();
    }, this.currentRound.duration * 1000);
  }

  replayHistory(session: ClientSession): void {
    for (const stroke of this.strokeHistory) {
      session.send(JSON.stringify(stroke));
    }
  }

  replayPlayers(session: ClientSession): void {
    for (const p of this.players.values()) {
      session.send(
        JSON.stringify({
          type: "join",
          payload: { id: p.id, username: p.username },
        }),
      );
    }
  }
}
